scripts/valhalla_build_extract.py

```diff
@@ -0,0 +1,157 @@
+
+description = """Builds a tar extract from the tiles in mjolnir.tile_dir to the path specified in mjolnir.tile_extract."""
+
+# "=" prefix means native byte order with standard size and no alignment:
+# https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment
+STRUCT_FORMAT = '=LQ'
+STRUCT_SIZE = struct.calcsize(STRUCT_FORMAT)
+INDEX_FILE = "index.bin"
+TAR_ARCNAME_DIR = Path('valhalla_tiles')
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-c", "--config", help="Absolute or relative path to the Valhalla config JSON.", type=Path)
+parser.add_argument("-v", "--verbosity", help="Accumulative verbosity flags; -v: INFO, -vv: DEBUG", action='count', default=0)
+
+# set up the logger basics
+LOGGER = logging.getLogger(__name__)
+handler = logging.StreamHandler()
+handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)5s: %(message)s"))
+LOGGER.addHandler(handler)
+
+
+def get_padded_tar_size(data_size: int) -> int:
+    """Returns the tar-internal file size, incl. header and NULL padding"""
+    #      <header>    <-data->    <-------------------NULL padding---------------->
+    return BLOCKSIZE + data_size + (BLOCKSIZE - (data_size % BLOCKSIZE or BLOCKSIZE))
+
+
+def get_tile_info(in_path: Path, tiles: dict):
+    """Recursively fills the passed map with tile path & size info"""
+    for p in in_path.iterdir():
+        if p.is_dir():
+            get_tile_info(p, tiles)
+        if p.is_file():
+            if not p.suffix == '.gph':
+                LOGGER.warning(f"Not a graph tile: {p}")
+                continue
+
+            tiles[p.resolve()] = get_padded_tar_size(p.stat().st_size)
+
+
+if __name__ == '__main__':
+    args = parser.parse_args()
+    with open(args.config) as f:
+        config = json.load(f)
+    index_fp: Path = args.config.parent.joinpath(INDEX_FILE)
+    extract_fp: Path = Path(config["mjolnir"]["tile_extract"])
+    tiles_fp: Path = Path(config["mjolnir"]["tile_dir"])
+
+    # set the right logger level
+    if args.verbosity == 0:
+        LOGGER.setLevel(logging.CRITICAL)
+    elif args.verbosity == 1:
+        LOGGER.setLevel(logging.INFO)
+    elif args.verbosity >= 2:
+        LOGGER.setLevel(logging.DEBUG)
+
+    if tiles_fp.exists() and tiles_fp.is_dir():
+        # collect the tile paths and file sizes
+        tiles_stats: Dict[Path, int] = dict()
+        get_tile_info(tiles_fp, tiles_stats)
+
+        print(tiles_stats)
+
+        index_size = STRUCT_SIZE * len(tiles_stats)
+        index_bin = b''
+        offset = get_padded_tar_size(index_size)  # initialize with the index file size inside the tar
+        for tile_path, tile_size in tiles_stats.items():
+            # extract level and ID portion to create a GraphId
+            # tile_path is the absolute path to the tile
+            tile_path = str(tile_path.relative_to(tiles_fp))[:-4]
+            level, idx = tile_path.split('/', 1)
+            tile_id = int(level) | (int(idx.replace('/', '')) << 3)
```

Comment:
currently this will throw if it's not parsing a valid tile filepath, even though we also only allow `.gph` files.. think I'll change that so also non-tiles can be tarred, the valhalla tar reader is also robust against invalid tile ids as tar names